src/librustc_privacy/lib.rs

```diff
@@ -898,190 +898,174 @@ impl<'a, 'tcx, 'v> Visitor<'v> for ObsoleteVisiblePrivateTypesVisitor<'a, 'tcx>
+        };
 
-                        if !vis.is_at_least(self.min_visibility, &self.tcx.map) {
-                            self.min_visibility = vis;
-                        }
-                        if !vis.is_at_least(self.required_visibility, &self.tcx.map) {
-                            if self.tcx.sess.features.borrow().pub_restricted ||
-                               self.old_error_set.contains(&ty.id) {
-                                let mut err = struct_span_err!(self.tcx.sess, ty.span, E0446,
-                                          "private type in public interface");
-                                err.span_label(ty.span, &format!("can't leak private type"));
-                                err.emit();
-                            } else {
-                                self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
-                                                       node_id,
-                                                       ty.span,
-                                                       format!("private type in public \
-                                                                interface (error E0446)"));
-                            }
-                        }
+        if let Some(def_id) = ty_def_id {
+            // Non-local means public (private items can't leave their crate, modulo bugs)
+            if let Some(node_id) = self.tcx.map.as_local_node_id(def_id) {
+                let item = self.tcx.map.expect_item(node_id);
+                let vis = ty::Visibility::from_hir(&item.vis, node_id, self.tcx);
+
+                if !vis.is_at_least(self.min_visibility, &self.tcx.map) {
+                    self.min_visibility = vis;
+                }
+                if !vis.is_at_least(self.required_visibility, &self.tcx.map) {
+                    if self.tcx.sess.features.borrow().pub_restricted || self.has_old_errors {
+                        let mut err = struct_span_err!(self.tcx.sess, self.span, E0446,
+                            "private type `{}` in public interface", ty);
+                        err.span_label(self.span, &format!("can't leak private type"));
+                        err.emit();
+                    } else {
+                        self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
+                                               node_id,
+                                               self.span,
+                                               format!("private type `{}` in public \
+                                                        interface (error E0446)", ty));
                     }
                 }
-                _ => {}
             }
         }
 
-        intravisit::walk_ty(self, ty);
+        if let ty::TyProjection(ref proj) = ty.sty {
+            // Avoid calling `visit_trait_ref` below on the trait,
+            // as we have already checked the trait itself above.
+            proj.trait_ref.super_visit_with(self)
```

Comment:
Yes on also the second point, `proj.super_visit_with(self)` would call `proj.trait_ref.visit_with(self)` which calls `self.visit_trait_ref(proj.trait_ref)` (which is not what we want).